solutions/gold/usaco-925.mdx

```diff
@@ -0,0 +1,127 @@
+---
+id: usaco-925
+source: USACO Platinum 2019 February
+title: Mooriokart
+author: Justin Ji
+---
+
+[Official Editorial (C++)](https://usaco.org/current/data/sol_mooriokart_platinum_feb19.html)
+
+<Spoiler title = "Hint 1">
+
+The graph is a forest, and we are only allowed to pick one path from each tree
+to reach our total. How can we write a knapsack DP to find our answer?
+
+</Spoiler>
+
+<Spoiler title = "Hint 2">
+
+The issue with directly writing a knapsack is that there is no bound on
+how long we can make our path. However, note how all paths of length $\ge Y$ can
+be compressed because they already reach the threshold.
+
+</Spoiler>
+
+<Spoiler title = "Explanation">
+
+We can first ignore the part about ordering the trees, and think of the problem
+from a knapsack DP perspective. Our goal is to choose a path from each tree that
+sums up to something $\ge Y$.
+
+For each tree, we can compress all the paths into $(length, frequency)$ pairs.
+Then, we can knapsack as usual. However, note that because we are compressing
+all paths of length $\ge Y$, we need to explicitly store the sum of the paths
+when writing our knapsack.
+
+Our DP state would be keep track of both the sum of the path lengths and
+the number of such paths, for each possible length of our path.
+
+Let $k$ be the number of trees in our forest. To account for the possible ways
+of ordering our paths, we need to multiply our initial sum of paths by
+$2^{k} * k!$. However, cyclic rotations and also reversed paths are considered
+to be the same, so we only multiply by $2^{k - 1} * (k - 1)!$.
+
+## Implementation
+
+**Time Complexity:** $\mathcal{O}(NY^{\frac{3}{2}})$
+
+<LanguageSection>
+<CPPSection>
+
+```cpp
+#include <bits/stdc++.h>
+using namespace std;
+using ll = long long;
+
+constexpr int MOD = 1e9 + 7;
```

Comment:
does this need to be initialized at compile time?

Reply:
i just write this out of habit lol, I'll change it to `const` later